app/src/main/java/org/simple/clinic/widgets/qrcodescanner/QrCodeScannerViewController.kt

```diff
@@ -0,0 +1,54 @@
+package org.simple.clinic.widgets.qrcodescanner
+
+import io.reactivex.Observable
+import io.reactivex.ObservableSource
+import io.reactivex.ObservableTransformer
+import io.reactivex.rxkotlin.Observables
+import io.reactivex.rxkotlin.ofType
+import org.simple.clinic.ReplayUntilScreenIsDestroyed
+import org.simple.clinic.ReportAnalyticsEvents
+import org.simple.clinic.util.RuntimePermissionResult
+import org.simple.clinic.widgets.ScreenCreated
+import org.simple.clinic.widgets.ScreenDestroyed
+import org.simple.clinic.widgets.TheActivityLifecycle
+import org.simple.clinic.widgets.UiEvent
+import javax.inject.Inject
+
+typealias Ui = QrCodeScannerView
+typealias UiChange = (Ui) -> Unit
+
+class QrCodeScannerViewController @Inject constructor() : ObservableTransformer<UiEvent, UiChange> {
+
+  override fun apply(events: Observable<UiEvent>): ObservableSource<UiChange> {
+    val replayedEvents = ReplayUntilScreenIsDestroyed(events)
+        .compose(ReportAnalyticsEvents())
+        .replay()
+
+    return Observable.merge(
+        startScanningForQrCode(replayedEvents),
+        stopScanningForQrCode(replayedEvents)
+    )
+  }
+
+  private fun startScanningForQrCode(events: Observable<UiEvent>): Observable<UiChange> {
+    val activityLifecycleEvents = events
+        .ofType<TheActivityLifecycle>()
+        .mergeWith(
+            events
+                .ofType<ScreenCreated>()
+                .map { TheActivityLifecycle.Resumed() })
+
+    return activityLifecycleEvents
+        .filter { lifecycleEvent -> lifecycleEvent is TheActivityLifecycle.Resumed }
```

Comment:
Do you want to do this instead?

```kotlin
activityLifecycleEvents
  .ofType<TheActivityLifecycle.Resumed>()
```

Reply:
I rewrote the chain using `.ofType()` 👍.